[4.0] Workflow Stage: Improve and fix delete and publish methods

chmst opened the pull request:
Pull Request for Issue #21846 .

### Summary of Changes
Remove the method delete() from stageTable to stageModel.
Doing this some fixes have been added, comments and code have been imprved


### Testing Instructions
Make a workflow with several stages and transitions. 
Now try to change the status of one or more stages, thrash them, unpublish them.
Watch alerts, are they correkt?
Use the filter "trashed" and delete trashed stages.
Make sure that the transitions to and from the deletes stages are deleted too.



### Expected result



### Actual result



### Documentation Changes Required

milux committed:
Tidy writeDynaList() (#12184)

* Cleaned writeDynaList() in core.js

* Removed explanation comments

* removed all API changes

* updated compressed core.js

phproberto committed:
[fix] publish/unpublish does not work with tables using null as default checked_out value (#20204)

mansourcodes committed:
Fix overwrite by .table-striped (#20180)

Fix overwrite by administrator/templates/isis/css/template.css line 1787

table.table-striped tbody > tr:nth-child(odd) > td,
table.table-striped tbody > tr:nth-child(odd) > th {
	background-color: #f9f9f9;
}

mansourcodes committed:
Fix overwrite by .table-striped (#20179)

Fix overwrite by administrator/templates/isis/css/template.css line 1787

table.table-striped tbody > tr:nth-child(odd) > td,
table.table-striped tbody > tr:nth-child(odd) > th {
	background-color: #f9f9f9;
}

brianteeman committed:
Tooltips not loading com_users (#20177)

The edit profile form is not loading the bootstrap tooltip code. So any tooltip (not popovers) are displayed as html as seen in the screenshot below when TFA is enabled.

This was spotted by @o2tsen and @sandewt while testing #20051 but as it is a bug effecting more than that PR I have created a new PR. (a pr should only fix one problem)

brianteeman committed:
[a11y] Headings consecutive order Debug Console (#20167)

> Headings communicate the organization of the content on the page. Web browsers, plug-ins, and assistive technologies can use them to provide in-page navigation.

> Skipping heading ranks can be confusing and should be avoided where possible: Make sure that a < h2> is not followed directly by an < h4>, for example.

Source (https://www.w3.org/WAI/tutorials/page-structure/headings/)

The headings were probably chosen for cosmetic reasons and not structural reasons which they should have been

This PR changes the heading in the debug console from h1 to h2

There is a very small visual change as a result but imho the benefits outweigh the small cost

brianteeman committed:
[a11y] Headings consecutive order (#20166)

* [WIP] [a11y] Headings consecutive order

> Headings communicate the organization of the content on the page. Web browsers, plug-ins, and assistive technologies can use them to provide in-page navigation.

> Skipping heading ranks can be confusing and should be avoided where possible: Make sure that a <h2> is not followed directly by an <h4>, for example.

Source (https://www.w3.org/WAI/tutorials/page-structure/headings/)

This PR changes the heading in the plugin and modules from h3 to h2 and in the template styles to h4

### todo
joomla.edit.item_title layout uses h4 but before I change it I need to check everywhere that it is being used

* layout

joomdonation committed:
Fix typos in InstallerControllerUpdate (#20154)

* Fix typos in InstallerControllerUpdate

* Fix same error on other places. Thanks @Quy

* Remove similar unnecessary code

* Revert "Remove similar unnecessary code"

This reverts commit 56410c0f59929d6a5afe220763fed6cfbcda6cdc.

* One more

* Revert "One more"

This reverts commit aa1b101f26be135b7161b4d71aa466f7556e5cab.

SharkyKZ committed:
[com_contact] Don't hide contact filter form (#20126)

* Update default_items.php

* Correct implode order.

* Codestyle

* More codestyle

SharkyKZ committed:
Fix for JUserHelper::addUserToGroup() when user group title is a number. (#20091)

* Update UserHelper.php

* Update UserHelper.php

Commit:
Fix count() in PHP 7.2 (#20044)

alikon committed:
[com_content][Multilanguage] - remove duplicated queries (#19683)

* [com_content][Multilanguage] - remove duplicated queries

* cs

* add $db->qn()

* removed ()

chmst committed:
Stage: Improve and fix delete and publish methods

chmst committed:
use casting for constants, use whereIn, fix empty &pks

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
 	 *
 	 * @param   array    &$pks   A list of the primary keys to change.
-	 * @param   integer  $value  The value of the published state.
+	 * @param   integer  $value     The target condition
```
Spacing

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
 	 * @since  __DEPLOY_VERSION__
 	 */
-	public function publish(&$pks, $value = 1)
+	public function publish(&$pks, $value = ContentComponent::CONDITION_PUBLISHED)
```
Da wird der Sniffer wahrscheinlich meckern, da das parent anders aussieht...

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
-		if ($value != 1)
+		$return = true;
+
+		if ($value != (int) ContentComponent::CONDITION_PUBLISHED)
```
```!==```

chmst replied:
!== funktioniert nicht, denn value ist int, die Konstante ist string. Wenn anstelle der Konstanten ein integer verwendet wird ist !== richtig

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+			// If the stage is trashed, the transitions to and from this stage must be trashed too
+			if ($return && $value == (int) ContentComponent::CONDITION_TRASHED)
+			{
+				$db = $this->getDbo();
```
```$db``` existiert schon oben?

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+				$db = $this->getDbo();
+				$query = $db->getQuery(true)
+					->update($db->quoteName('#__workflow_transitions'))
+					->set($db->quoteName('published') . ' = ' . ContentComponent::CONDITION_TRASHED)
```
Hier darf nicht die Artikel-Condition genommen werden sondern tatsächlich ```-2```

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+			self::checkDefaultStage($pks, $value);
+		}
+
+		if (!empty($pks))
```
```$pks``` sollte auf int geparst werden ```ArrayHelper::toInteger```

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+			$return =  parent::publish($pks, $value);
+
+			// If the stage is trashed, the transitions to and from this stage must be trashed too
+			if ($return && $value == (int) ContentComponent::CONDITION_TRASHED)
```
```-2``` statt Konstante

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+		$app = Factory::getApplication();
+
+		// Clear pks
+		self::checkDefaultStage($pks, ContentComponent::CONDITION_TRASHED);
```
```-2``` statt Konstante

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+			$query = $db->getQuery(true)
+				->delete($db->quoteName('#__workflow_transitions'))
+				->where($db->quoteName('to_stage_id') . ' = ' . (int) $pk, 'OR')
+				->whereIn($db->quoteName('from_stage_id'), $pks);
```
Das stimmt glaub nicht, das ```(int) $pk``` müsste wahrscheinlich auch ein IN(...) sein?

chmst replied:
done

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+		$query = $db->getQuery(true)
+			->select($db->quoteName(array('id')))
+			->from($db->quoteName('#__workflow_stages'))
+			->where($db->quoteName('default') . ' =1 ')
```
Spacing (rechts)

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+	public function checkDefaultStage(&$pks, $value)
+	{
+		$db  = $this->getDbo();
+		$pks   = (array) $pks;
```
Auf ```INT```parsen...

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
+
+		if (!empty($default))
+		{
+			if ($value == (int) ContentComponent::CONDITION_TRASHED)
```
Keine Konstante

bembelimen commented on administrator/components/com_workflow/Model/StageModel.php:
```diff
@@ -292,36 +293,132 @@ public function setDefault($pk, $value = 1)
-				if ($table->load(array('id' => $pk)) && $table->default)
+				$app->enqueueMessage(Text::_('COM_WORKFLOW_MSG_DELETE_DEFAULT'), 'error');
+			}
+			elseif ($value == (int) ContentComponent::CONDITION_UNPUBLISHED)
```
Keine Konstante

chmst committed:
Merge branch '4.0-dev' into workflow-stage-delete-publish

chmst committed:
see Review

chmst committed:
make sure $pks js an array of integers

chmst committed:
align the "="

chmst committed:
satisfy sniffer

chmst committed:
Fix spaces

wilsonge committed:
Update custom elements to new version

laoneo committed:
[4.0] Remove deprecated item title layout (#22678)

* Remove deprecated item title layout

* Remove usages

laoneo committed:
[4.0] Introduce cache controller service provider (#22687)

* Introduce cache controller service provider

* change key

* add log triggers

* Introduce CacheControllerInterface

* change return type

* Default to output controller

* Revert "Introduce CacheControllerInterface"

This reverts commit 89940aa

* revert cache controllerinterface

laoneo committed:
[4.0] Module dispatcher (#19834)

wilsonge committed:
Merge commit 'e05d733' into 4.0-dev

wilsonge committed:
Merge commit '0b349eb4a1820b6fcd564c5796e2ca7a868baa9d' into 4.0-dev

brianteeman committed:
[4.0] Sidebar Menu [a11y] (#22692)

* a11y

* typo

chmst committed:
Merge branch '4.0-dev' into workflow-stage-delete-publish

wilsonge committed:
Fix contact and newsfeeds to the newly named classes

wilsonge committed:
Convert sidebar colours to be CSS Variables

wilsonge committed:
Fix checkout icons in list views

brianteeman committed:
[4.0] Update build script (#22706)

### The problem
when running the scripts with #22263 they were trying to process the choices css files as if they were js.
this was only present on windows

### The solution??
hopefully this is correct. It works for me but anything with regex and js and I am stabbing in the dark

laoneo committed:
Alpha order imports of com_admin

laoneo committed:
Alpha order imports of com_associations

laoneo committed:
Alpha order imports of com_banners

laoneo committed:
Alpha order imports of com_ajax

laoneo committed:
Alpha order imports of com_cache

laoneo committed:
Alpha order imports of com_categories

laoneo committed:
Alpha order imports of com_checkin

laoneo committed:
Alpha order imports of com_config

laoneo committed:
Alpha order imports of com_contact

laoneo committed:
Alpha order imports of com_content

laoneo committed:
Alpha order imports of com_contenthistory

laoneo committed:
Alpha order imports of com_cpanel/tmpl

laoneo committed:
Alpha order imports of com_csp

laoneo committed:
Alpha order imports of com_fields

laoneo committed:
Alpha order imports of com_finder

laoneo committed:
Alpha order imports of com_installer

laoneo committed:
Alpha order imports of com_joomlaupdate

laoneo committed:
Alpha order imports of com_languages

laoneo committed:
Alpha order imports of com_login

laoneo committed:
Alpha order imports of com_media

laoneo committed:
Alpha order imports of com_menus

laoneo committed:
Alpha order imports of com_messages

laoneo committed:
Alpha order imports of com_mailto

laoneo committed:
Alpha order imports of com_modules

laoneo committed:
Alpha order imports of com_newsfeeds

laoneo committed:
Alpha order imports of com_plugins

laoneo committed:
Alpha order imports of com_postinstall

laoneo committed:
Alpha order imports of com_redirect

laoneo committed:
Alpha order imports of com_search

laoneo committed:
Alpha order imports of com_tags

laoneo committed:
Alpha order imports of com_templates

laoneo committed:
Alpha order imports of com_users

laoneo committed:
Alpha order imports of com_workflow

laoneo committed:
Alpha order imports of com_wrapper

laoneo committed:
Fix the banners categories state lookup

laoneo committed:
New line after namespace imports

laoneo committed:
Alpha order imports of mod_quickicon module

laoneo committed:
Remove not needed imports in blog style layout

laoneo committed:
Fix code style errors

chmst committed:
resolve conflict

chmst committed:
Update StageModel.php

chmst committed:
Merge branch '4.0-dev' into workflow-stage-delete-publish
